Add support for detecting docker container

ashu-mehra opened the pull request:
Add support in port library for detecting if we are running in
docker container or not.
This is required for cgroup related APIs in port library,
such as `omrsysinfo_cgroup_get_memlimit`, to work properly.
This is because when running in a docker container,
the cgroups listed in `/proc/self/cgroup` do not match the cgroups
of the process in the container. `/proc/self/cgroup` actually
shows the cgroups as seen from the host.

Eg if a process is running in a docker container,
its `/proc/PID/cgroup` on docker will show something like:
```
$ cat /proc/8364/cgroup
11:cpuset:/docker/<container id>
10:perf_event:/docker/<container id>
9:memory:/dockera/<container id>
...
```

but the actual cgroup of the process would be "/".

Once we figure out that we are running in docker container,
then cgroup APIs in port library should use "/" as cgroup name
for reading resource limits.

Note that as and when docker containers start supporting
cgroup namespace, then above mentioned mismatch would also be fixed.
Following issues and PRs are related to adding support for
cgroup namespace in docker container:
https://github.com/opencontainers/runc/pull/1184
https://github.com/opencontainers/runtime-spec/issues/62
https://github.com/opencontainers/runtime-spec/issues/66

There is no direct mechanism to detect if we are running in
docker container.
Most reliable mechanism that I have come across is
by checking the cgroup for PID 1 (by reading `/proc/1/cgroup` file)
for all subsystems. If all cgroups are "/", then it indicates the
process is running on host system.
Else we conclude it to be running in a docker container.
Note that this mechanism will break when docker containers start
supporting cgroup namespace.

Signed-off-by: Ashutosh Mehra <asmehra1@in.ibm.com>

ashu-mehra commented:
@DanHeidinga can you please review this.

ashu-mehra commented:
Sorry, last minute changes resulted in those failures. I have fixed them. It should pass now.

charliegracie commented:
genie-omr build all

ashu-mehra commented:
@DanHeidinga can you please review this.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3603,6 +3650,56 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+		 * For any other cgroup name, assume we are in a 'docker' container.
+		 */
+		rc = readCgroupFile(portLibrary, 1, FALSE, &cgEntryList, NULL);
+		if (0 != rc) {
```
The `goto` is unnecessary if the `!=` is changed to `==` and the code below is moved into the if block.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3613,25 +3710,23 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
+			omrthread_monitor_enter(cgroupEntryListMonitor);
+			if (NULL == PPG_cgroupEntryList) {
+				if (OMR_CONTAINER_DOCKER == PPG_containerType) {
+					rc = readCgroupFile(portLibrary, getpid(), TRUE, &PPG_cgroupEntryList, &PPG_cgroupSubsystemsAvailable);
```
A local variable for the `TRUE` vs `FALSE` would be clearer rather than duplicating the `readCgroupFile` call

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3613,25 +3710,23 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
-		omrthread_monitor_exit(cgroupEntryListMonitor);
-		if (0 != rc) {
-			goto _end;
+		if (isCgroupV1Available(portLibrary)) {
```
Do I read this code correctly that the cgroup detection is only possible on cgroup v1 systems?

ashu-mehra replied:
For cgroup v2, the file system type mounted on `/sys/fs/cgroup` is different. The function `isCgroupV1Available()` is currently only handling cgroup v1, therefore I named it so.
As and when we add support for cgroup v2, I believe we can rename the function then.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3603,6 +3650,56 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+	}
+_end:
+	if (NULL != cgEntryList) {
+		freeCgroupEntries(portLibrary, cgEntryList);
```
This seems allocation heavy for the information being gathered.  Can this be done without allocating and freeing the cgroup list?

ashu-mehra replied:
It looks possible. Let me see if I can refactor it to make it light weight.

ashu-mehra commented on include_core/omrport.h:
```diff
@@ -1012,6 +1012,13 @@ typedef struct OMROSKernelInfo {
 
+typedef enum OMRContainerType {
+	OMR_CONTAINER_NONE,
+	OMR_CONTAINER_DOCKER,
```
I checked `runc` container and it displays same behavior with respect to `/proc/pid/cgroup` file. So I believe the detection mechanism should work for any `runc`-style container, not just Docker.
Since Docker is most popular system at the moment, I named it so.
Do you want to rename it as something like `OMR_CONTAINER_RUNC`?

ashu-mehra commented on port/unix/omrsysinfo.c:
```diff
@@ -3431,6 +3472,7 @@ readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, OMRCgroupEntry **cgr
 	requiredSize = portLibrary->str_printf(portLibrary, NULL, (uint32_t)-1, "/proc/%d/cgroup", pid);
 	Assert_PRT_true(requiredSize <= PATH_MAX);
 	portLibrary->str_printf(portLibrary, cgroup, sizeof(cgroup), "/proc/%d/cgroup", pid);
+
```
`cgroup` at this point is storing path `/proc/pid/cgroup`. Same var is also used for storing cgroup name by parsing `/proc/pid/cgroup` file. This may be the cause of confusion. I will update code to rename the variable.
Use of `ROOT_CGROUP` is intentional as we don't want to use the cgroup name we get from `/proc/pid/cgroup` file.

ashu-mehra commented on port/unix/omrsysinfo.c:
```diff
@@ -3735,6 +3831,48 @@ omrsysinfo_cgroup_get_memlimit(struct OMRPortLibrary *portLibrary, uint64_t *lim
+
+#if defined(LINUX) && !defined(OMRZTPF)
+	{
+		const char *envValue = getenv("CONTAINER_TYPE");
```
Its a fallback mechanism for current detection logic. As I mentioned in the commit message, container detection logic is flaky and can break in future. If that happens, then we can use this env variable to indicate to runtime that its running in a container.
For example, we can set this env variable in Dockerfile to enable container awareness in the runtime.

ashu-mehra commented:
> I'm not sure I follow the need for (2). Can you clarify why these apis are useful and what port library consumers would do with this info?

I thought it may come handy if runtimes want to display information about the container they are running in for diagnostic purpose. For example, OpenJ9 can display the container type in javacore?

ashu-mehra commented:
@DanHeidinga I have updated the code as per the comments. Can you please take a look.

ashu-mehra commented:
> This is a little hard to review as it mixes together two things:
> 1. fixing a bug(?) around detecting and handling non-v1 cgroup support

I should point out that the change is not to handle non-v1 cgroup support. Changes are mainly to  make cgroup support work properly in (docker) container by detecting that we are in the container environment.

ashu-mehra commented:
@DanHeidinga can you please review the updated change set.

mstoodle commented:
another ping for @DanHeidinga who maybe missed the first one due to conference circuit prep distractions :) .

DanHeidinga commented:
Thanks @mstoodle and @ashu-mehra - My goal is to have reviewed this by end of week (hold me to it!)

DanHeidinga commented on include_core/omrport.h:
```diff
@@ -1463,6 +1471,10 @@ typedef struct OMRPortLibrary {
+	/** see @ref omrsysinfo_get_container_type "omrsysinfo_get_container_type"*/
+	int32_t ( *sysinfo_get_container_type)(struct OMRPortLibrary *portLibrary, OMRContainerType *containerType);
+	/** see @ref omrsysinfo_get_container_name "omrsysinfo_get_container_name"*/
+	const char * ( *sysinfo_get_container_name)(struct OMRPortLibrary *portLibrary, OMRContainerType containerType);
```
These two APIs don't belong in this changeset.  Given the VM only supports detecting linux cgroups or nothing, we don't have enough examples to generalize to a good portable api.  

I would rather forgo these APIs now and add them when we have second (or third) container system to work with.  That will result in a better designed api that is easier to maintain longterm.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -256,6 +256,18 @@ struct  {
 #define OMR_CGROUP_V1_MOUNT_POINT "/sys/fs/cgroup"
+#define ROOT_CGROUP "/"
+
+/* Following is the description of /proc/<pid>/cgroup copied from 'man' page for proc:
```
We can't copy from the `man` page for proc as the documentation is under GPLv2: http://man7.org/linux/man-pages/man5/proc.5.license.html

ashu-mehra replied:
I updated the comment to put it in my own words instead of copying from `man` page.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -299,6 +311,12 @@ struct {
 	{ "memory", OMR_CGROUP_SUBSYSTEM_MEMORY }
 };
 
+const char *OMRContainerNames[OMR_CONTAINER_NUM_CONTAINERS] = {
```
As per comment above, this belongs in a separate changeset or should wait until we have more examples of container systems to build the api on.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3300,6 +3320,37 @@ omrsysinfo_os_kernel_info(struct OMRPortLibrary *portLibrary, struct OMROSKernel
+		goto _end;
+	}
+
+_end:
```
label can be removed.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3603,6 +3652,69 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+ * @return 0 on success, otherwise negative error code
+ */
+static int32_t
+getContainerType(struct OMRPortLibrary *portLibrary, OMRContainerType *containerType)
```
Please remove this method from this PR.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3613,25 +3725,24 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
-		if (0 != rc) {
-			goto _end;
+				if (OMR_CONTAINER_DOCKER == PPG_containerType) {
+					useRootCgroup = TRUE;
```
This will need to be reworked when the container type api is moved out.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3706,13 +3817,14 @@ int32_t
 {
 	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM;
+
+	Assert_PRT_true(NULL != limit);
```
Please update the function prototype to indicate `limit` cannot be null.

ashu-mehra replied:
Done

ashu-mehra commented:
@DanHeidinga I have updated the change set to remove `sysinfo_get_container_type` and `sysinfo_get_container_name` APIs.
Other comments are also addressed. Please review the updated change set.

DanHeidinga commented on port/common/omrsysinfo.c:
```diff
@@ -902,6 +902,7 @@ omrsysinfo_cgroup_are_subsystems_enabled(struct OMRPortLibrary *portLibrary, uin
  * omrsysinfo_cgroup_enable_limits() before calling this function.
  * When the fuction returns OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM,
  * value of *limits is unspecified.
+ * Note that 'limit' parameter should not be NULL.
```
Nitpick: `should not` -> `must not`  

This is a bit of a nitpick but `must` is a stronger requirement than `should`.  Since the code will assert it is not null, `must` is a better choice.

DanHeidinga commented on port/win32/omrsysinfo.c:
```diff
@@ -1653,3 +1653,4 @@ omrsysinfo_cgroup_get_memlimit(struct OMRPortLibrary *portLibrary, uint64_t *lim
 {
 	return OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM;
 }
+
```
Can you remove this file from the PR?  It doesn't have any real changes.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3444,34 +3483,28 @@ readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, OMRCgroupEntry **cgr
 
 	while (0 == feof(cgroupFile)) {
+		char cgroup[PATH_MAX];
+		char subsystems[1024];
```
Can you add a comment that points to the `subsystemNames` array to indicate why `1024` is likely sufficient?

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3617,6 +3655,63 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+				break;
+			} else if (3 != rc) {
+				rc = portLibrary->error_set_last_error_with_message_format(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED, "unexpected format of /proc/1/cgroup file");
+				goto _end;
```
`@return 0 on success, otherwise negative error code`

This may exit the method with a positive, non-zero return value contrary to the function contract.

ashu-mehra replied:
`rc` is reassigned return value of `error_set_last_error_with_message_format` which is negative error code.

DanHeidinga replied:
I missed that - sorry.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3617,6 +3655,63 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+				goto _end;
+			}
+
+			if (strcmp(ROOT_CGROUP, cgroup)) {
```
Can you make this a boolean comparison and explicitly add the `!= 0` check to the if statement?  The port library (or more specifically, new code) tries to ensure that all comparisons result in boolean values.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3617,6 +3655,63 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+ * @return 0 on success, otherwise negative error code
+ */
+static int32_t
+isRunningInDocker(struct OMRPortLibrary *portLibrary, BOOLEAN *inDocker)
```
Should this be `isRunningInContainer`?  It appears to generically check for a container runtime, not docker specifically.

ashu-mehra replied:
`isRunningInContainer` is also fine. The mechanism works for runc based containers, not specific to docker.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3617,6 +3655,63 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
+ * Checks if the process is running inside docker container 
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[out] inDocker pointer to BOOLEAN which on successful return indicates if the process is running in docker container or not 
```
Why not return `TRUE` if running in a container and `FALSE` for not / errors?  

The current approach means user code needs to check two things:
```
  BOOLEAN inDocker = false;
   if (0 == isRunningInDocker(portLib, &inDocker)) {
      if (inDocker == TRUE) {
        /* take action */
      }
    }
```
Is there a meaningful distinction between error and not in docker?

ashu-mehra replied:
I am inclined to make this distinction between error and not in docker.
Its caller `omrsysinfo_cgroup_is_system_available` should not attempt to do any processing if any error is encountered.
If we treat error case same as 'not in docker' case, then we may end up in trouble if the error happened when its actually running in docker. In such case it would attempt to read non-existing cgroup control files for determining resource limits.
To be on safer side, if any error occurs, better to not use cgroup limits at all and fallback to regular behavior.

DanHeidinga commented on port/unix/omrsysinfo.c:
```diff
@@ -3627,25 +3722,26 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
+			}
+			omrthread_monitor_enter(cgroupEntryListMonitor);
+			if (NULL == PPG_cgroupEntryList) {
+				BOOLEAN useRootCgroup = FALSE;
```
Is `useRootCgroup` required or can the code pass `inDocker` instead?

ashu-mehra replied:
We can use `inDocker` instead.

ashu-mehra commented:
@DanHeidinga change set is updated. Please review.

DanHeidinga commented:
@ashu-mehra I've applied these changes to an OpenJ9 build and added some debug prints but I'm not seeing any output.  Is there a commandline flag / env var I need to set to enable cgroup detection?

ashu-mehra commented on port/unix/omrsysinfo.c:
```diff
@@ -3627,25 +3722,21 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
-		omrthread_monitor_exit(cgroupEntryListMonitor);
-		if (0 != rc) {
-			goto _end;
+			rc = isRunningInContainer(portLibrary, &inContainer);
```
I noticed that but didn't bother as `isCgroupV1Available` just adds an overhead of one syscall.
But if we want to common it out, how about calling the new function `checkCgroupPrereqs(&cgroupVersion, &inContainer)` where `cgroupVersion` would be an `enum 
 { V1, V2 }` (currently the code only supports V1, but sooner than later we would have to consider V2 as well).

DanHeidinga replied:
Since we'll have to touch this code again to add V2 support, I'm OK with leaving it as is now.

ashu-mehra commented on port/unix/omrsysinfo.c:
```diff
@@ -3627,25 +3722,21 @@ omrsysinfo_cgroup_is_system_available(struct OMRPortLibrary *portLibrary)
-			goto _end;
+			rc = isRunningInContainer(portLibrary, &inContainer);
+			if (0 != rc) {
+				goto _end;
```
>  we should capture some diagnostics to know why we failed to determine if we were in a container or not.

I agree. I actually want to add tracepoints in the code at various error conditions and thought of taking that up it in a separate PR. Is that ok?

DanHeidinga replied:
Works for me.

charliegracie commented:
@genie-omr build all zos
